### ISSUE TITLE
Feature Request: Alternative Estimation of Multilevel Reliability 

### ISSUE THREAD

**Issue by weoweigelt** [simsem/semTools#106]
Dear semTools-developers,

I very much appreciate the great features semTools offers. The reliability-function is so user friendly especially when it comes to multilevel data. Recently, I learned that the procedures described by Geldhof and colleagues (2014) to estimate multilevel relaibility may produce biased estimates for level 2 reliability. There is a recent paper by Lai (2021) providing some code for R. 

There is a discussion on researchgate:
https://www.researchgate.net/post/How_do_I_estimate_reliability_for_multilevel_data_in_R

I wondered whether it would be possible to integrate the improved procedures suggested by Lai into the reliability function of semTools, e.g., as an additional estimate besides the classic one? If there is any chance this would be great and much appreciated by any lay user of R like me. ESM researchers run multilevel CFA by default. Adding one line of code (reliability(my.fitted.mode) is so handy. 

Best
Oliver

Geldhof, G. J., Preacher, K. J., & Zyphur, M. J. (2014). Reliability estimation in a multilevel confirmatory factor analysis framework. Psychological Methods, 19(1), 72–91. https://dx.doi.org/10.1037/a0032138
Lai, M. H. C. (2021). Composite reliability of multilevel data: It’s about observed scores and construct meanings. Psychological Methods, 26(1), 90–102. https://doi.org/10.1037/met0000287

**Comment by TDJorgensen** [comment]
Thanks, I'm aware of Lai's article and I agree with him.  I'm thinking about how best to incorporate these into `reliability()`.  The function was already a bit sloppy to refer to 3 different omegas, and now this would be even more of them.  I would like to add arguments that instead request model-implied vs. observed total variance in the denominator (instead of printing both `omega2` and `omega3`), as well as requesting composite reliability vs. reliability of latent scores.  That could apply not only to Geldhof et al.'s alpha and omegas, but also to "ordinal alpha" for categorical indicators, which apply only to the latent response variables rather than observed data (i.e., no need to print `alpha` and `alpha_ord`).  I'm even thinking about an argument for assuming tau-equivalence, so it simply returns alpha when `TRUE` or omega when `FALSE`.

I also read a convincing case for not using a single symbol omega for all the various reliability estimators derived from factor models.  But I fear that ship has somewhat sailed.

**Comment by marklhc** [comment]
A colleague shared this thread with me, and I appreciate the semTools developers for considering adding the indices discussed in my paper. For your information, since the paper was published, I have made a function [here](https://github.com/marklhc/mcfa_reliability_supp/blob/master/multilevel_alpha.R) for computing the alpha and omega indices, but I haven't done a lot of testing with it. I'll be thrilled to provide support for incorporating the indices into the `semTools` package. Thank you for making and maintaining this wonderful tool.

Best,

Mark

**Comment by TDJorgensen** [comment]
Thanks for sharing your software @marklhc.  I have made a few commits the past few days related to this.  First, I put `AVE()` in its own function, because it is not a `reliability()` coefficient anyway:

- https://github.com/simsem/semTools/commit/b84731e9ea4cc742010026f59dd6e6296703cc9b

Then I created a function `compRelSEM()` to calculate composite reliability coefficients from SEM estimates. 

- https://github.com/simsem/semTools/commit/13df89237deb142f5d2103581626e30e81536f02

It will replace `reliability()` (soon to be deprecated), and includes several new arguments for the user to target their desired type of coefficient (e.g., alpha by setting argument `tau.eq=TRUE`), rather than returning several (which can be confusing, and possibly promote choosing the largest coefficient to report, rather than the one with the appropriate interpretation).  Details available on the help page, if you install the development version:

```
devtools::install_github("simsem/semTools/semTools")
```

Geldhof et al.'s (2014) are still the default for MLSEMs because the user must specify which constructs are configural or shared in order to obtain any of Lai's (2021) coefficients (in which case Geldhof et al.'s will not be returned).  Construct types are specified with either (or both) arguments:

- `config=` (99383de) returns coefficients appropriate for Level-1 constructs (`_W` and `_2L`)
- `shared=` (6424e8e) returns coefficients appropriate for Level-2 constructs (`_B` and an IRR coefficient I define in an upcoming paper)

Before I send a new version to CRAN, I would be interested in getting feedback from @weoweigelt and @marklhc.  And if @marklhc compares my output to his own repository's R output, make sure to set `obs.var=FALSE` in order to use the model-implied total variance in the denominator (i.e., functions of Lambda and Theta) instead of the "observed" variance (functions of estimated Sigmas from saturated models).  Note that I implemented multidimensional generalizations of the formulas proposed in Lai (2021), consistent with `reliability()`'s `omega2` (`obs.var=FALSE`) or `omega3` (`obs.var=TRUE`, the default), which also simplifies the process of choosing the variance to include in the denominator.

Thanks for any feedback!

**Comment by marklhc** [comment]
@TDJorgensen, thanks for the amazing work! I've started comparing the results and documenting the process [here](https://github.com/marklhc/mcfa_reliability_supp/blob/master/compare_semTools.md). I used the example in Lai (2021). As you can see, for the configural model, the results are essentially the same (other than the rounding error for my calculation of cluster size). I actually did not realize `lavaan` can allow using the same names for the latent factors at different levels, and it seems that `compRelSEM()` depends on it, so maybe good to mention in the documentation.

The requirement to have the same names for the latent factors may explain the discrepancy in (a) a model with a saturated within-cluster component and (b) a simultaneous shared-and-configural model by Stapleton & Johnson (2019). It seems that when computing reliability for shared constructs, `compRelSEM()` currently does not take into account the level-1 variance when there's not a latent factor with the same name at level 1. See the [last example](https://github.com/marklhc/mcfa_reliability_supp/blob/master/compare_semTools.md#simultaneous-shared-and-configural-model=). I see the documentation suggests not using a saturated within-cluster component, so would it be helpful to give an error or a warning when the within-component is not found? On the other hand, the simultaneous shared-and-configural model was discussed in Jak et al. (2021), and right now `compRelSEM()` seems to use different denominators for the configural and the shared constructs at level 2.

I will continue testing the tau equivalence part in a few days. Thanks again and really excited to see this gets implemented (together with IRR).

**Comment by weoweigelt** [comment]
@TDJorgensen thank you creating these new functions and for sharing them with us. I have tested the compRelSEM-function and I have tried different settings. The reliabilities estimated with the improved procedures look plausible. For instance, reliability of a scale drops from .90 (Geldhof-style) to .87 (Lai-style: tau.eq=TRUE, config = all factors). 

Admittedly, being mostly a cookbook SEM-user, I have yet to familiarize myself with the implications of the different settings. For instance, tau.eq=FALSE returns a warning that loadings are not equal across levels for one of the scales/factors. This is a very clear and helpful warning. I appreciate this and I appreciate that the Omegas are estimated anyway. I am kind of impressed that semTools examines tau equivalence explcitly. 

Upon playing around with the settings estimated coefficients did not differ much beginning with the third decimal place. This probably may not generalize to other data, but is kind of comforting to me. It seems I may not necessarily draw completely different/wrong conclusions even if assumptions are violated. 

My initial impulse was that I would prefer a default setting that "does the right thing", but I understand that this is absolutely non-sensical. Working with SEM and analyses like these requires a minimum of familiarity with the underlying assumptions and the type of construct depends on the specific research context and variables considered. 

If I can be of help by doing an acid test from the perspective of a beginner, feel free to let me know. Again, thank you for providing this incredible function.

**Comment by TDJorgensen** [comment]
Thank you both for the feedback!

@marklhc 

> I actually did not realize `lavaan` can allow using the same names for the latent factors at different levels, and it seems that `compRelSEM()` depends on it, so maybe good to mention in the documentation.

Indeed, I rely on the same name across levels in order to match the indicators used in the composite (e.g., to check for equal loadings, issuing a warning when they are unequal).  I have updated the documentation of the `config=` argument to note that the same name is expected, and I also added a warning when a factor name is not detected on both levels of analysis: [e71e57c](https://github.com/simsem/semTools/commit/e71e57ce1894c0c54ad7ed13d81f5378ebe16f5f)

You are correct that this reliance on matching factors explains the discrepancies you noted with the saturated within-level model, as well as the simultaneous shared-and-configural model.  In those cases, no indicator names were identified at Level 1, so no subset of Sigma-within was extracted to represent within-level variance.  I have fixed this in the development version by first checking for empty Level-1 names in the case of a `shared=` construct, in which case I now substitute with Level-2 indicator names that are among the Level-1 names (just in case there are additional indicators that are at Level 2 only).  Thus, shared-construct reliability no longer requires fitting a configural model, despite my (colleague's) hesitations about the appropriateness of such a model, and the documentation is updated accordingly.

@weoweigelt 

> `tau.eq=FALSE` returns a warning that loadings are not equal across levels for one of the scales/factors. This is a very clear and helpful warning. I appreciate this and I appreciate that the Omegas are estimated anyway. I am kind of impressed that semTools examines tau equivalence explicitly.

It sounds like you are confusing invariance (equal loadings across levels) with tau-equivalence (equal loadings across indicators of the same factor).  I do check for invariance of configural-factor loadings, and warn of detected violations.  I do not check the tau-equivalence assumption because it is not necessary.  When `tau.eq=TRUE`, coefficient alpha is calculated using the observed (or model-implied, when `obs.var=FALSE`) covariance matrix, which does not assume equal loadings across indicators.

> I would prefer a default setting that "does the right thing", but I understand that this is absolutely non-sensical. Working with SEM and analyses like these requires a minimum of familiarity with the underlying assumptions and the type of construct depends on the specific research context and variables considered.

Indeed, I tried to choose the most reasonable default settings.  From a practical standpoint of implementation, it is simply impossible to know what kind of multilevel coefficients to calculate without the user telling me via the `config=` and/or `shared=` arguments, so Geldhof et al.'s (2014) coefficients have to remain the default unless more information is provided.  I suppose I could calculate every possible coefficient, but that is a behavior of the deprecated `reliability()` function that I was hoping to avoid with `compRelSEM()`.
